### PR TITLE
Update to v23.2.0-IOFreeze

### LINKED ISSUE #4934: Update to EnergyPlus 23.2.0
Enhancement Request
--------------
Once v23.2-IOFreeze is released, we should update to it.

## Detailed Description

https://github.com/NREL/EnergyPlus/blob/develop/src/Transition/InputRulesFiles/Rules23-1-0-to-23-2-0.md
https://github.com/NREL/EnergyPlus/blob/develop/src/Transition/OutputRulesFiles/OutputChanges23-1-0-to-23-2-0.md

### Major changes

- [x] Fuel Type enum changes (`OtherEquipment`, `Exterior:FuelEquipment`, `ZoneHVAC:HybridUnitaryHVAC`, etc.)
  - "DistrictHeating" to "DistrictHeatingWater"
  - "Steam" to "DistrictHeatingSteam"
  - https://github.com/NREL/OpenStudio/pull/4968
  - [x] Watch status of this PR and add DistrictCooling/DistictHeatingSteam in LCC objects
      - https://github.com/NREL/EnergyPlus/pull/10229

- #4958 
    - [x] `Coil:Cooling:DX:*`, `Coil:Heating:DX:*`, `Coil:WaterHeating:*`: new "Crankcase Heater Capacity Function of Temperature Curve Name" field
    - [x] `Coil:Cooling:DX:TwoSpeed`, `Coil:Cooling:DX:VariableSpeed` `Coil:Heating:DX:VariableSpeed`: add fields "High Speed 2017 Rated Evaporator Fan Power Per Volume Flow Rate", "High Speed 2023 Rated Evaporator Fan Power Per Volume Flow Rate", "Low Speed 2017 Rated Evaporator Fan Power Per Volume Flow Rate", and "Low Speed 2023 Rated Evaporator Fan Power Per Volume Flow Rate"
    - [x] `AirLoopHVAC:UnitarySystem`, `AirLoopHVAC:UnitaryHeatPump
:WaterToAir` (not wrapped), `ZoneHVAC:WaterToAirHeatPump`: remove a handful of fields
    - [x] `Coil:Cooling:DX:VariableSpeed`, `Coil:Cooling:WaterToAirHeatPump:VariableSpeedEquationFit`: add fields "Maximum Cycling Rate", "Latent Capacity Time Constant", "Fan Delay Time"
    - [x] `Coil:Cooling:WaterToAirHeatPump:ParameterEstimation` (not wrapped), `Coil:Cooling:WaterToAirHeatPump:EquationFit`: add fields "Part Load Fraction Correlation Curve Name", "Maximum Cycling Rate", "Latent Capacity Time Constant", "Fan Delay Time"
   - [x] `Coil:Heating:WaterToAirHeatPump:ParameterEstimation` (not wrapped), `Coil:Heating:WaterToAirHeatPump:EquationFit`: add field "Part Load Fraction Correlation Curve Name"

- [x] `HeatPump:PlantLoop:EIR:Cooling`, `HeatPump:PlantLoop:EIR:Heating`: required field changes, and some new fields
  - https://github.com/NREL/OpenStudio/pull/4962

### New additions
- [ ] `ZoneEarthtube`: two new fields; `ZoneEarthtube:Parameters`: new object
  - https://github.com/NREL/OpenStudio/issues/4943
- [ ] `SpaceHVAC:EquipmentConnections`, `SpaceHVAC:ZoneEquipmentSplitter`, `SpaceHVAC:ZoneEquipmentMixer`: new object
  - https://github.com/NREL/OpenStudio/issues/4944
- [x] `DistrictHeating:Steam`: new object
  - https://github.com/NREL/OpenStudio/issues/4941
  - https://github.com/NREL/OpenStudio/pull/4954
  - https://github.com/NREL/OpenStudio/pull/4972
  - https://github.com/NREL/OpenStudio-resources/pull/200
- [ ] `PlantEquipmentOperation:ChillerHeaterChangeover`: new object
  - https://github.com/NREL/OpenStudio/issues/4942

### Minor stuff
- [x] `DistrictHeating` to `DistrictHeating:Water`: new object name in forward translation
- [x] `SurfaceProperty:LocalEnvironment`: change "External Shading Fraction Schedule Name" to "Sunlit Fraction Schedule Name"
- [x] `HVACTemplate:Zone:WaterToAirHeatPump` (not wrapped): rename "Heat Pump Time Constant" to "Latent Capacity Time Constant" and remove "Fraction of On-Cycle Power Use"
- [x] `ZoneHVAC:EvaporativeCoolerUnit` (not wrapped): add field "Shut Off Relative Humidity"
- [ ] `ZoneHVAC:WaterToAirHeatPump`, `ZoneHVAC:TerminalUnit:VariableRefrigerantFlow`: add fields "Design Specification Multispeed Object Type" and "Design Specification Multispeed Object Name"
  - https://github.com/NREL/OpenStudio/pull/4964
- [x] `Coil:Heating:Fuel`, `Coil:Heating:Gas:MultiStage`, `Coil:Heating:Desuperheater`: rename "Parasitic Electric Load" to "On Cycle Parasitic Electric Load" and "Parasitic Fuel Load" to "Off Cycle Parasitic Fuel Load"
  - https://github.com/NREL/OpenStudio/pull/4970
- [x] `Boiler:HotWater`: rename "Parasitic Electric Load" to "On Cycle Parasitic Electric Load" and add "Off Cycle Parasitic Fuel Load"
  - https://github.com/NREL/OpenStudio/pull/4949
  - https://github.com/NREL/OpenStudio/pull/4985
- [x] `Controller:OutdoorAir`: new field "Economizer Operation Staging"
  - https://github.com/NREL/OpenStudio/pull/4948
- [x] `LoadProfile:Plant`: new fields "Plant Loop Fluid Type", "Degree of SubCooling", "Degree of Loop SubCooling"
  - https://github.com/NREL/OpenStudio/pull/4951
- [x] `EnvironmentalImpactFactors`: update field names

### Other
- [x] Fix a bunch of spelling in ProposedEnergy+.idd, including exectuted, repeate, similiar, etc.
- [x] `ConstructionProperty:InternalHeatSource`: remove `\note required-field` line
- [x] `Construction:ComplexFenestrationState`: fix "Absoptance" spelling in fields

### PR DESCRIPTION
**Pull request by joseph-robertson**
Pull request overview
---------------------

 - Fixes #4934

### Pull Request Author

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

**Comment by joseph-robertson** [comment]
How do we want to handle changes from "DistrictHeating" -> "DistrictHeatingWater" and "Steam" -> "DistrictHeatingSteam"? Should we:
- A) Leave the idd intact and silently convert in FT
- B) Add the new enums to the idd, leave and deprecated the old ones (how?), and do conversion in FT along with a warning
- C) break backward compatibility and just straight up change the idd

@jmarrec

**Comment by jmarrec** [comment]
@joseph-robertson we want to update the IDD for sure. And catch it at model time ideally.

I'm not 100% sure this is going to work flawlessly, but here's what I'm thinking: these should always have been `setFuelType(openstudio::FuelType fuelType)` ones. I suspect this was impossible at the time because E+ wasn't consistent enough, but I think it could be now.

We want to modify the `FuelType` enum to support both here: https://github.com/NREL/OpenStudio/blob/0ab02805c62a5f459e669bf3f4c65c95a1d8eed7/src/utilities/data/DataEnums.hpp#L64-L82

```diff
+ // When there are two entries, the first one is an alias. The second one, .valueDescription(), is the one E+ uses
OPENSTUDIO_ENUM(FuelType,
[...]
  ((DistrictCooling))
- ((DistrictHeating))
- ((DistrictHeating)(DistrictHeatingWater))
+ ((DistrictHeatingSteam))
  ((OtherFuel_1)(OtherFuel1))
```

The FuelType could be created from both `DistrictHeating` and `DistrictHeatingWater`, and we would consistently use `.valueDescription()` for interfacing with E+.


Ideally, any of these classes with setFuelType(std:string) should have two methods

```c++
bool setFuelType(openstudio::FuelType fuelType) {
   return setString(index, fuelType.valueDescription());   // This catches the case where the IDD only supports a subset of the fuel types
}

bool setFuelType(const std::string& fuelType) {   // We could mark this one as deprecated too, so we stop supporting them soonish, but I think it's helpful to have the ability to do both
    try {
       openstudio::FuelType fuelTypeEnum(fuelType);  // Allows creating from both
       return setFuelType(fuelTypeEnum);
    } catch (std::runtime_error& e) {
       LOG(Warn, e.what());
      return false;
   }
}
```

File: resources/energyplus/ProposedEnergy+.idd [hunk 1]
```diff
@@ -32556,9 +32887,9 @@ Coil:Heating:DX:VariableSpeed,
         \memo (includes electric compressor and outdoor fan), variable-speed, with defrost
         \memo controls. Requires two to ten sets of performance data and will interpolate between
         \memo speeds.
-   \extensible:7
-   \min-fields 25
-   \max-fields 88
+        \extensible:7
```
Comment (jmarrec) [inline resources/energyplus/ProposedEnergy+.idd:32890]: This should be bumped to 9. I will do that on the Coils PR, but just a note that there could be other occurrences where bumping extensible was needed.

File: resources/energyplus/ProposedEnergy+.idd [hunk 2]
```diff
@@ -31006,7 +31274,7 @@ Coil:Cooling:DX:VariableSpeed,
         \memo ten sets of performance data and will interpolate between speeds. Modeled as a
         \memo single coil with variable-speed compressor.
         \extensible:10
-        \min-fields 31
+        \min-fields 35
```
Comment (jmarrec) [inline resources/energyplus/ProposedEnergy+.idd:31277]: Same here, extensible needs to be bumped to 12

**Review by jmarrec — COMMENTED** [review]
We need to be careful about bumping `\extensible:` in the IDD. The FT does HORRIBLE things if you don't do it. I think there were only two objects receiving and extensible fields change. Will correct it on the Coils PR (I wouldn't have caught it if I didn't had a ton of FT tests at your request @joseph-robertson !)

**Comment by jmarrec** [comment]
Ok, time to drop the hammer. This builds and tests fine. Merging to develop, this will simplify our dev workflow and allow us to be more reactive and run reg tests more easily.

There's an E+ RC1 out, might become official, but there may be an RC2 as well.

**Pull request merged**